nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java

```diff
@@ -0,0 +1,180 @@
+    @WritesAttribute(attribute = "aws.kinesis.firehose.record.id", description = "Record id of the message posted to Kinesis Firehose")})
+public class PutKinesisFirehose extends AbstractKinesisFirehoseProcessor {
+
+    /**
+     * Kinesis put record response error message
+     */
+    public static final String AWS_KINESIS_FIREHOSE_ERROR_MESSAGE = "aws.kinesis.firehose.error.message";
+
+    /**
+     * Kinesis put record response error code
+     */
+    public static final String AWS_KINESIS_FIREHOSE_ERROR_CODE = "aws.kinesis.firehose.error.code";
+
+    /**
+     * Kinesis put record response record id
+     */
+    public static final String AWS_KINESIS_FIREHOSE_RECORD_ID = "aws.kinesis.firehose.record.id";
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME, BATCH_SIZE, MAX_MESSAGE_BUFFER_SIZE_MB, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE, AWS_CREDENTIALS_PROVIDER_SERVICE, TIMEOUT,
+                  PROXY_HOST,PROXY_HOST_PORT));
+
+    /**
+     * Max buffer size 1000kb
+     */
+    public static final int MAX_MESSAGE_SIZE = 1000 * 1024;
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        FlowFile flowFileCandidate = session.get();
+        if ( flowFileCandidate == null )
+            return;
+
+        long currentBufferSizeBytes = flowFileCandidate.getSize();
+
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final int maxBufferSizeBytes = context.getProperty(MAX_MESSAGE_BUFFER_SIZE_MB).asInteger() * 1024 * 1000;
+
+        // Get max batch size messages with size limit defined by maxBufferSizeBytes
+        List<FlowFile> flowFiles = new ArrayList<FlowFile>(batchSize);
+        flowFiles.add(flowFileCandidate);
+        for (int i = 1; (i < batchSize) && (currentBufferSizeBytes <= maxBufferSizeBytes); i++) {
+            flowFileCandidate = session.get();
+            if ( flowFileCandidate == null )
+                break;
+            currentBufferSizeBytes += flowFileCandidate.getSize();
+            flowFiles.add(flowFileCandidate);
+        }
+
+        final String firehoseStreamName = context.getProperty(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME).getValue();
+        final AmazonKinesisFirehoseClient client = getClient();
+
+        try {
+            List<Record> records = new ArrayList<>();
+
+            List<FlowFile> failedFlowFiles = new ArrayList<>();
+            List<FlowFile> successfulFlowFiles = new ArrayList<>();
+
+            // Prepare batch of records
+            for (int i = 0; i < flowFiles.size(); i++) {
+                FlowFile flowFile = flowFiles.get(i);
+
+                // If flow file too large send it to failure
```

Comment:
On second thought, I am okay with this.  Seeing as how the Firehose also batches/aggregates I do not feel there is any issue with sending batch sizes that may be a bit smaller